doc/design/determinism.md

```diff
@@ -0,0 +1,107 @@
+# Determinism in iceoryx
+
+Deterministic exectution is important for many application domains, especially real-time systems.
+
+However, the meaning of determinism differs between domains such as
+computer science and real-time systems.
+
+## Deterministic result
+
+The function result only depends on the input and is entirely predictable in the sense that repeated
+execution with the same input (and state of the environment) always leads to the same
+output.
+
+The state of the environment refers to any global or member variables the function depends on.
+In particular it may depend on the current time or a pseudo random number generator state.
+
+This is the (informal) definition used in computer science.
+
+## Deterministic runtime
+
+The runtime of a function is bounded and the function cannot block indefinitely under normal
```

Comment:
What are normal conditions?

Reply:
Good question, I will remove it or specify it. I think it might be helpful to allow infinite blocking in abnormal conditions, which for me means a system failure.

Say we trigger a fatal error, from then on all bets are of and the function will possibly not return. This is also a design philosophy in our error handling (currently and will be extended). We call terminate, but the function does not return. Would this be called blocking? Also, it is not guaranteed that termination happens "immediately" etc.

There is also the possibility to send the system into an infinite loop during FATAL errors (some design approaches do this, as kind of a safe state). This would also be possible in the future.

TLDR: I would like to state that during FATAL errors no runtime bound can be assumed.

Reply:
Clarified?